src/base/emu-i386/simx86/trees.c

```diff
@@ -1276,8 +1276,9 @@ static void do_invalidate(unsigned data, int cnt)
 	cnt = PAGE_ALIGN(data+cnt-1) - (data & PAGE_MASK);
 	data &= PAGE_MASK;
 #ifdef HOST_ARCH_X86
-	if (!CONFIG_CPUSIM)
-		InvalidateNodeRange(data, cnt, 0);
+	/* see comment below in e_invalidate() below */
+	assert(!CONFIG_CPUSIM || (CEmuStat & CeS_INSTREMU));
```

Comment:
Would it be possible to not change
CONFIG_CPUSIM and only add a
flag check when interp chooses what
back-end to run? Or is it a too big change?

Reply:
I could change CONFIG_CPUSIM but not config.cpusim? Is that what you mean?
Actually almost all (except for some parser.y debug output) use of CONFIG_CPUSIM outside simx86 is of the form:
`IS_EMU() && !CONFIG_CPUSIM`
so probably better to introduce
`IS_EMU_JIT()` ?
for that. Then CONFIG_CPUSIM can move from cpu-emu.h to emu86.h and check for instremu  or regular sim.

Reply:
IS_EMU_JIT() is ok, but I mean something
else. So that you don't change config.cpusim
and not change CONFIG_CPUSIM, so that
all code still thinks its JIT, so pages are kept
invalidated on writes, and so on.
And just add an extra flag check in a place
where interp choses  what back-end to run.

Reply:
I.e. running sim back-end when config.cpusim
still set, doesn't look like a problem. Or does it?

Reply:
Then it would be better to have those places not use CONFIG_CPUSIM any more (probably in another PR then).

Most uses of CONFIG_CPUSIM are in interp.c and cpu-emu.c and need the interpretive branch for instremu (flagsync etc), so it would be a bigger change to change all those.

What remains is in trees.c/sigsegv.c. Some of those aren't necessary, e.g. `e_vgaemu_fault` is only called for `InCompiledCode`.

Reply:
And in either case, changing CONFIG_CPUSIM
looks better than changing config.cpusim.

Reply:
> I could change CONFIG_CPUSIM but not config.cpusim?

At least. :)

Reply:
Resolved with #1982